### ISSUE TITLE
Niagara sometimes replaces an app in overview

### ISSUE THREAD

**Issue by MathisP75** [8bitPit/Niagara-Issues#2571]
### Describe the bug

While using gesture navigation in my Pixel, sometimes, going into overview will result in the Niagara Launcher app taking the place of the last opened app. This makes the app impossible to access, as clicking on it in the overview opens the home screen.

### Steps to reproduce

I currently do not know the steps to reproduce this bug, but I will be searching for a pattern whenever it occurs in the future. I also do not know if this happens with other 3rd party launchers, I will try using the other ones I have to verify. I will update this issue if I find anything relevant.

### Does the issue also occur with other 3rd-party launchers?

Not applicable

### Device name

Google Pixel 7 Pro

### Android version

Android 14

### Screenshots / Screen recordings

![Screenshot_20240124-230449](https://github.com/8bitPit/Niagara-Issues/assets/98901170/c16abd61-f3eb-4789-aa33-e1b301a4302f)

**Comment by MathisP75** [comment]
I managed to get a [screen recording of it happening](https://photos.app.goo.gl/vJbKZ5QS6DEPqqMd7). The bug occurs near the end of it.

**Comment by NiagaraSupport** [comment]
Hey there, and thanks for the report.
We're aware of several gesture navigation issues on Pixel phones right now. We'll look into it on our end, but it might also prove helpful to address this on [Google's issue tracker](https://issuetracker.google.com/issues?q=third%20party%20launcher%20gestures%20recent%20apps) as it seems to extend beyond our control.

**Comment by milanglacier** [comment]
I am thinking about an alternative method to bypass the "gesture" problem.

Based on my usage, it is clear that such problem only happens in the homescreen. I also note that if opening the "recent apps" by calling the API rather than from gesture (e.g. some assistive touch app), then the "recent apps" displays normally.

Attached is a video to show the difference between opening "recent apps" via API or via gesture.

https://github.com/8bitPit/Niagara-Issues/assets/45728125/c529fe7d-39cb-4127-9dd2-48a44d1613ad


In Niagara, we can swipe up to open "search", and if we can use "swipe up" to call the API to open the "recent apps", and allow user to choose the behaviors (either open search panel or "recent apps") for swiping up, then this situation will be much more mitigated.

**Comment by TFackler** [comment]
@milanglacier 
Thank you for the suggestion. 

However, we think the proposed changes do not fit with our design philosophy. Adding a toggle that changes the swipe-up gesture to do multiple different things could lead to confusion and less intuitive usage. On top of that, it would only be helpful for a fraction of users, namely the ones experiencing this issue.

If we were to add these changes, we should also add other workarounds for bugs other users are experiencing. This would lead to a lot of toggles for functionality that only some users need, which could complicate everything further.

**Comment by gusano** [comment]
While I totally understand that fixing a bug for a device model is not ideal, Niagara has become unusable on my Pixel 6 Pro because of this big and that's a bummer because Pixel phones are widely used 😬
I am not a fan of the above proposal either but just saying...

**Comment by MathisP75** [comment]
I agree that this issue should really be looked into more, as it hinders a lot the experience.

**Comment by TFackler** [comment]
We have been able to reproduce the issue on other third-party launchers and have come to the conclusion that this is an Android issue rather than just a Niagara Launcher issue.

We still do not feel like implementing our own version of the recent apps screen is warranted.

**Comment by MathisP75** [comment]
> We have been able to reproduce the issue on other third-party launchers

Ah well, that sucks... I should test more launchers, but I always come back to Niagara Launcher quickly.

**Comment by gusano** [comment]
@TFackler IIUC I have the feeling that the discussion is drifting towards a "recent apps" issue when the original issue is about icons being replaced and Niagara becoming unresponsive. 
I have that same issue on my Pixel 6 Pro and I disabled the often used apps option.
I will upload a screen recording later.

**Comment by gusano** [comment]
https://github.com/user-attachments/assets/4058ec99-b7f2-48e0-b134-d2fc876d9fbf

**Comment by MathisP75** [comment]
> @TFackler IIUC I have the feeling that the discussion is drifting towards a "recent apps" issue when the original issue is about icons being replaced and Niagara becoming unresponsive. 
> I have that same issue on my Pixel 6 Pro and I disabled the often used apps option.
> I don't have any issue with others launchers.
> I will upload a screen recording later.
This issue is not about what you posted in your screen recording. Refer to the one I posted earlier to see what I am referring to.

**Comment by itpeters** [comment]
I believe I have a 100% (or close to it) reproduction of this behavior. There is a small difference between the original report and what I am capturing. I see a buggy entry in my overview that has the correct app icon (as in the original), but the image displayed is off my background without any elements of the Niagara home screen.

Perhaps most interestingly, the reproduction steps _only_ work in Niagara and not in other third party launchers (tested in Nova and Lawnchair). The step that reliably triggers this bug in Niagara does not cause issues in other third party launchers. All tests of each launcher were done with that launcher set as the default home app, tested on my Pixel 9 Pro XL running the January stable update (AP4A.250105.002), with Niagara 1.13.3.

The crucial step for me is to swipe up from the bottom as if performing the home gesture, but slowly enough to feel the haptic feedback which indicates that releasing the gesture would trigger the recent app overview. Once you feel that feedback, do not release the gesture, but instead return your thumb to the bottom of the screen, allowing the app to remain in the foreground.

Once you've done this, if you swipe home, and then perform the gesture to open the recent app overview, you will see the bugged entry in the recents overview. Interestingly, this broken state can survive other interactions. If instead of returning home you perform a sideways swipe on the pill to switch to a different app, swipe home from that app, and then trigger the recent app overview from the Niagara home screen, you will still see the bug. Conversely, if you perform the repro step, use the pill to switch from one foreground app to another, and then use the pill to return to the original app, you will not see the broken state in the recent apps overview.

Interacting with the broken app from the overview (the app which displays as the home screen background) has varying behavior, sometimes requiring more than one tap to get back to the app, and seems to lead to less deterministic behavior. Opening the broken app from the Niagara home screen, rather than the recents overview, reliably clears the broken state and gets recents working correctly again.

I have made a recording of interacting with my phone (with show touches on screen enabled) and triggering this bug as described above with 100% reliability. I do not believe that this is the only issue confronting third party launchers and the recent apps functionality, but it seems valuable in that (a) it is a reliable reproduction and (b) this particular pathway only seems to affect Niagara launcher. These steps work without issue on every other third party launcher I have tested.

https://photos.app.goo.gl/XuLeJoTk8H5x4z686

**Comment by MathisP75** [comment]
@itpeters This is exactly it, I can reproduce. I only see my wallpaper now (unlike in the original bug report, so something must have changed since), but it's 100% the same bug.

I was never able to figure out what exactly caused the issue, but it seems like you just did!

Let's hope they can find a solution now.

**Comment by itpeters** [comment]
@MathisP75 glad that the reproduction works for you as well! I am learning that I like to fidget with apps without swiping home in a way that tends to reproduce this constantly without meaning to.

Here are a few more observations I've found:

1. There are clear visual differences that something odd is happening even before you get to the strange wallpaper-only entry in the recents overview. Without the bug, the Niagara home screen can be seen scaling down and turning translucent, eventually transparent, as the recents gesture is completed, with the correct most recent app visible to the left; with the bug triggered, the home screen doesn't change at all and the wrong app is visible to the left.
2. My reproduction case can be simplified somewhat. It is enough to open an app, use the gesture to open the recent app overview from that app, tap the app that was open (returning it to the foreground), use the home gesture to go to the Niagara home screen, and then use the recent app overview gesture again.
3. I have tested a few more third party launchers, and I have found others that have this same behavior, using the exact same reproduction steps. This suggests to me that there must be something in the details of how a launcher works with the available Android APIs.
      1. Lawnchair (tested 14.Dev.(d7e3f32) does not trigger this bug.
      2. Microsoft launcher (tested 6.241202.1.1162300) **does** trigger this bug!
      3. Nova launcher (tested 8.0.18) does not trigger this bug.
      4. Smart Launcher (tested 6.5 build 004) does not trigger this bug.
      5. Action Launcher (tested 51.0) **does** trigger this bug!

@NiagaraSupport would it be possible to have someone on the team take another look at this issue in light of straightforward reproduction steps, along with evidence that not every third party launcher runs into this? I'm a happy and paying Niagara user, but this issue is easily my most frustrating cosmetic issue -- plus, given that it makes interacting with the recent app overview inconsistent, it's not just cosmetic.

**Comment by itpeters** [comment]
I just added three short videos to the album link I shared above (https://photos.app.goo.gl/XuLeJoTk8H5x4z686) showing the repro steps on each of Niagara and Microsoft Launcher (bugs out) and Nova Launcher (no bug) for completeness.

**Comment by MathisP75** [comment]
I have found that Kvaesitso and AIO Launcher also trigger this bug

**Comment by itpeters** [comment]
Other launchers that have the same bug:

- KISS Launcher
- Olauncher

Launchers that do **not** have the same bug:

- hyperion launcher
- Square Home

**Comment by itpeters** [comment]
I found another reproduction method that works for me all of the time. Launch an app, hold down the power button to invoke Gemini (my configured assistant), click anywhere on the screen that is not the Gemini panel to close Gemini, then swipe home from the app. After that sequence, invoking the overview from the home screen will have this bug present for the app that had been open.

I haven't tested every launcher we tested today, but I tried a few, and so far if the earlier reproduction triggered the bug then this one will as well. That tentatively makes me think that there is a common root cause, and that whatever the launchers that don't suffer from this bug are doing will resolve this issue via whichever path you arrive at it.

In terms of that common element -- this is hand-wavy, but it feels like having the focus removed from the foreground app without having focus return to the launcher (via the overview screen, via the Gemini panel) is enough to get things broken. I wonder if there are other apps that draw over other apps that can be invoked without closing the foreground app to see if they could trigger this same behavior.

EDIT: After pondering for a bit, I thought of another thing that steals focus from the foreground app. Chat bubbles! I managed to recall how to create a chat bubble in Messages, and then experimented with Niagara and Nova as examples of launchers that do and do not have this bug. Just having the chat bubble enabled doesn't trigger the bug, but opening and then closing the chat bubble while another app is in the foreground, swiping home, and then pulling up the recent app overview recreates the bug just like the other reproductions identified. Once again, the bug is present in Niagara and not in Nova.

I wonder if there's any back channel communication between third party launcher developers where the Niagara folks could bring this up.  I'd love to know what Nova and Smart Launcher are doing (or not doing!) that means they aren't susceptible to this.

(I added four more videos to the shared album, showing Niagara and Nova handling the Gemini popup and the message bubble, still at https://photos.app.goo.gl/XuLeJoTk8H5x4z686).

**Comment by itpeters** [comment]
I verified again this morning that current versions of Lawnchair Launcher do not trigger this bug using any of the reproduction steps identified yesterday. Since that project is open source, I was hoping to find something in the source or relatively recent commit history that seemed like it could be related. Unfortunately, despite a number of commits in the upstream Android launcher code that seemed enticing, nothing held up.

I also tried using older versions of Lawnchair Launcher to see if I could bisect when any change might have gone in. Here's what I found:

- As mentioned yesterday, the current released version of Lawnchair (`14.Dev.(d7e3f32)`) doesn't reproduce this bug.
- The most recent "released" version of Lawnchair, `14.Beta 3` from August 2024, also doesn't have this bug.
- Lawnchair `12.1.0 Alpha 1` (May 2022) was the first version of Lawnchair based on Launcher3 from Android 12.1, and it does not have this bug.
- The last version of Lawnchair before it was based on Launcher3 was `1.2.0.1884` from all the way back in July 2018. Despite a warning on install about being built for a very old SDK version, it does run, and more importantly, it exactly reproduces the bug being discussed using the same reproduction steps.

From all of this, I'm hypothesizing that Niagara was based on an older AOSP launcher and is missing potentially important updates, including changes that would fix how Niagara interacts with the recent app overview.

I'm going to shamelessly tag @TFackler, who has been active in recent Niagara issues, but I think I'm at my limit, between my lack of experience developing for Android and working from the outside of the Niagara code base. We have reliable reproductions, demonstrated that this is not an inherent limitation of third-party launchers in general, and found at least circumstantial evidence that this behavior stems from older launcher code bases and more modern Android versions.

Hopefully, speaking as a big fan of Niagara, there's enough here to look into and hopefully resolve this!

**Comment by itpeters** [comment]
OK, back with another update. I got curious how Lawnchair handles the recent app overview (just continuing to work with Lawnchair because the source is available with a long history spanning when it did and then later did not have this bug).

Just grepping around I ran into the `RecentsGestureHandler` which actually seems like it was intended to be a Lawnchair option for configuring certain gestures on the launcher itself (e.g. swipe down on the home screen to open the notifications shade) but never got wired in. That said, it does show a way of invoking the recent app overview.

https://github.com/LawnchairLauncher/lawnchair/blob/15-dev/lawnchair/src/app/lawnchair/gestures/handlers/RecentsGestureHandler.kt

The immediately interesting code there for me was:

```kt
        app.performGlobalAction(AccessibilityService.GLOBAL_ACTION_RECENTS)
```

On a hunch I decided to check Tasker to see if it exposed a way of triggering the recents overview as a task without needing to spin up my own development environment, and sure enough it does (using the accessibility service). I created a Tasker widget to call this task, added it to my Niagara home, and gave it a shot.

I'll add a new video to the album (https://photos.app.goo.gl/XuLeJoTk8H5x4z686) but the summary is that invoking recents this way:

1. includes a blank, as in truly blank, app as the rightmost entry in the overview, and clicking on it returns to the Niagara home screen
2. correctly shows the recent apps, in the correct order, with no graphical issues, to the left of this totally blank entry, regardless of whether Niagara has gotten into a bugged state or not.

To expand on that second point, if I perform any of the repro steps identified above, Niagara will display a buggy app entry in the recents overview when using the home pill gesture to bring up the recents, but clicking on the Tasker widget to bring up the recent apps still works correctly.

I'm not quite sure what to make of this yet. It's very intriguing to me that there is a basically correct recent app overview there and then there's something very different when triggering it via gesture when using Niagara. I'll also note that the method `performGlobalAction` is found in https://github.com/LawnchairLauncher/lawnchair/blame/15-dev/lawnchair/src/app/lawnchair/LawnchairApp.kt and there's some other intriguing looking (and per git, original to the Launcher3 refactor of Lawnchair) code in there about lifecycle management and the recents overview.

**Comment by itpeters** [comment]
I believe I have gotten to the bottom of this. Or, at least, figured out what the difference is between launchers that do and do not have this bug.

I set Android Studio up, found the simplest open source launcher that I could find which demonstrates this bug and were easy to import into Android Studio (KISS Launcher). I was able to quickly get it building locally and, when deployed to my phone, verified that the debug build was still showing this issue.

Next, I created a new project using the "Empty Views Activity" template, modified the manifest to add `android.intent.category.HOME` and `android.intent.category.DEFAULT` to the single activity, and verified that I could now set my new app on my phone to be the world's most useless launcher.

At this point my useless launcher did not demonstrate the bug using any of the reproductions from above, but my local build of KISS did. At this point I started trying a combination of stripping things out of KISS and adding things to my useless launcher to see if I could figure out the difference.

I'll skip a ton of false starts (target SDK versions, event handlers, etc.) and cut to the chase. Adding the `android:excludeFromRecents="true"` attribute to the activity in my useless launcher causes it to suffer from the bug we're talking about. Conversely, removing that attribute from my build of KISS launcher resolves the bug.

Importantly, not setting the attribute while having the application set as the default launcher still correctly does not show it in the recents overview _when it is set as the default home app_, so specifying it seems to be both unnecessary and to trigger an Android bug (almost like the system is trying to remove it from recents twice, or by two different pathways).

TLDR: Remove the `android:excludeFromRecents="true"` from the `AndroidManifest.xml` for Niagara to resolve this bug.

**Comment by Maxr1998** [comment]
Thank you so much for digging into this and finding a root cause for this issue! And even more for the detailed write-up. We have now applied the change internally and it will be part of the next update.

<details>
<summary>Some more technical background on this issue, for those interested.</summary>

From what I've seen, the bug only occurs on Pixel devices (or at least devices using Launcher3's quickstep implementation, which Pixel Launcher is based on) since Android 14. I can also reproduce it on my Pixel 5, but never saw it on my Samsung phones.
Generally, applying the `excludeFromRecents` flag shouldn't cause any issues, but there seems to be a bug in Android/Launcher3, as you suggested.
I did some additional investigation and found a potential trigger in the [`TopTaskTraker.CachedTaskInfo.getVisibleNonExcludedTask()`](https://android.googlesource.com/platform/packages/apps/Launcher3/+/refs/heads/main/quickstep/src/com/android/quickstep/TopTaskTracker.java#245) in Launcher3. It's possible that marking Niagara's home activity as `excludeFromRecents` causes this method to not return `null` for a task group containing Niagara Launcher, and ultimately causes the observed bug. There are checks for the home app in this method, but they might not apply for other reasons. I'd have to add some logging/attach a debugger to this method, but unfortunately debugging framework code is a little more complex and requires compiling your own system image (or otherwise patching the system code). Thus, I won't look closer into this, but if anyone reading this is wants to verify this theory, I'd be intrigued to hear the results.
</details>

**Comment by itpeters** [comment]
@Maxr1998 Fantastic, so glad to hear it. I also appreciate you digging a bit deeper, since it's fairly obviously a bug in the underlying system that shouldn't be triggerable by other applications. It's lucky that there's a workaround from the outside that (fingers crossed!) doesn't have any negative side effects.

The class you identified seems really promising. My rough thinking is that the repro steps all involve additional activities that are themselves excluded from recents, and there's some logical corner case where the existing checks are insufficient for the non-trivial case. That said, I am a burnt out/retired(?) dev, and this investigation was already closer to coding again than I've tried to let myself get. I was starting to feel some guilt for not continuing to pursue it, but hearing what is involved in debugging framework code makes it easier to set it down here!

I suppose trying to open an issue in an appropriate Android bug tracker is another option, but it feels like it would be simpler to just punch myself in the face and cut out the intermediate steps.

@MathisP75 That's a good thought. Perhaps just linking them to this issue with a brief note and let them make of it what they will.

**Comment by Maxr1998** [comment]
> Fantastic, so glad to hear it. I also appreciate you digging a bit deeper, since it's fairly obviously a bug in the underlying system that shouldn't be triggerable by other applications. It's lucky that there's a workaround from the outside that (fingers crossed!) doesn't have any negative side effects.

Hopefully! We'll roll it out as a beta first, if there are any issues we should receive feedback from our users.

> The class you identified seems really promising. My rough thinking is that the repro steps all involve additional activities that are themselves excluded from recents, and there's some logical corner case where the existing checks are insufficient for the non-trivial case.

Right, debugging and fixing this issue would probably require a lot of effort in advance just to set up everything. Instead, we'll probably do a write up for Google and let their engineers fix this bug.

> That said, I am a burnt out/retired(?) dev, and this investigation was already closer to coding again than I've tried to let myself get. I was starting to feel some guilt for not continuing to pursue it, but hearing what is involved in debugging framework code makes it easier to set it down here!

Oh, no worries, there's no need to feel guilty. Your contributions were already a lot more than what I could have dreamed of. Thanks again!

> I suppose trying to open an issue in an appropriate Android bug tracker is another option, but it feels like it would be simpler to just punch myself in the face and cut out the intermediate steps.

There's already an issue for this bug [here](https://issuetracker.google.com/issues/350745502), but it's unfortunately market as wontfix. I'll nonetheless copy our investigation there so that Google is aware.

> That's a good thought. Perhaps just linking them to this issue with a brief note and let them make of it what they will.

I can also handle that, we're internally in contact with other launcher developers already.
